crates/services/txpool_v2/src/service.rs

```diff
@@ -1,234 +1,157 @@
+pub type Service<View> = ServiceRunner<Task<View>>;
 
-        let view = self
-            .persistent_storage_provider
-            .latest_view()
-            .map_err(|e| Error::Database(e.to_string()))?;
+pub struct SelectTransactionsRequest {
+    pub constraints: Constraints,
+    pub response_channel: oneshot::Sender<Vec<ArcPoolTx>>,
+}
 
-        for transaction in transactions {
-            self.heavy_async_processor.spawn({
-                let shared_state = self.clone();
-                let params = params.clone();
-                let view = view.clone();
-                async move {
-                    // TODO: Return the error in the status update channel (see: https://github.com/FuelLabs/fuel-core/issues/2185)
-                    let checked_tx = perform_all_verifications(
-                        transaction,
-                        shared_state.pool.clone(),
-                        current_height,
-                        &params,
-                        version,
-                        shared_state.gas_price_provider.as_ref(),
-                        shared_state.wasm_checker.as_ref(),
-                        shared_state.memory.get_memory().await,
-                        view,
-                    )
-                    .await
-                    .unwrap();
-                    #[allow(unused_variables)]
-                    let result = {
-                        let mut pool = shared_state.pool.write();
-                        // TODO: Return the result of the insertion (see: https://github.com/FuelLabs/fuel-core/issues/2185)
-                        pool.insert(checked_tx)
-                    };
-                }
-            });
-        }
-        Ok(())
-    }
+pub enum WritePoolRequest {
+    InsertTxs {
+        transactions: Vec<Arc<Transaction>>,
+    },
+    InsertTx {
+        transaction: Arc<Transaction>,
+        response_channel: oneshot::Sender<Result<(), Error>>,
+    },
+    RemoveCoinDependents {
+        transactions: Vec<(TxId, String)>,
+    },
 }
 
-pub type Service<
-    PSProvider,
-    ConsensusParamsProvider,
-    GasPriceProvider,
-    WasmChecker,
-    MemoryPool,
-> = ServiceRunner<
-    Task<PSProvider, ConsensusParamsProvider, GasPriceProvider, WasmChecker, MemoryPool>,
->;
+pub enum ReadPoolRequest {
+    GetTxIds {
+        max_txs: usize,
+        response_channel: oneshot::Sender<Vec<TxId>>,
+    },
+    GetTxs {
+        tx_ids: Vec<TxId>,
+        response_channel: oneshot::Sender<Vec<Option<ArcPoolTx>>>,
+    },
+}
 
-pub struct Task<
-    PSProvider,
-    ConsensusParamsProvider,
-    GasPriceProvider,
-    WasmChecker,
-    MemoryPool,
-> {
-    shared_state: SharedState<
-        PSProvider,
-        ConsensusParamsProvider,
-        GasPriceProvider,
-        WasmChecker,
-        MemoryPool,
-    >,
+pub struct Task<View> {
+    chain_id: ChainId,
+    subscriptions: Subscriptions,
+    verification: Verification<View>,
+    p2p: Arc<dyn P2PRequests>,
```

Comment:
Do we need to hold a trait object here? Can't we use a generic and save a heap allocation?

```suggestion
pub struct Task<View. P2PRequests> {
    chain_id: ChainId,
    subscriptions: Subscriptions,
    verification: Verification<View>,
    p2p: P2PRequests,
```

Reply:
The choice has been made to reduce the generics and use the Arc even if it's less performant because they are mainly used in tasks that consume a lot and so they will not be our bottleneck. + we win code readability

Reply:
Makes sense :+1: